abr_control/controllers/osc.py

```diff
@@ -56,164 +54,214 @@ def __init__(self, robot_config, kp=1, ko=None, kv=None, ki=0, vmax=None,
-        Mx_inv = np.dot(J, np.dot(M_inv, J.T))
-        if abs(np.linalg.det(Mx_inv)) >= 1e-5:
-            # do the linalg inverse if matrix is non-singular
-            # because it's faster and more accurate
-            Mx = np.linalg.inv(Mx_inv)
-        else:
-            # using the rcond to set singular values < thresh to 0
-            # singular values < (rcond * max(singular_values)) set to 0
-            Mx = np.linalg.pinv(Mx_inv, rcond=.005)
+        M = self.robot_config.M(q)  # inertia matrix in joint space
+        Mx, M_inv = self._Mx(M=M, J=J)  # inertia matrix in task space
 
         # calculate the desired task space forces -----------------------------
         u_task = np.zeros(6)
 
-        # calculate position error if position is being controlled
+        # if position is being controlled
         if np.sum(self.ctrlr_dof[:3]) > 0:
-            # calculate the end-effector position information
             xyz = self.robot_config.Tx(ref_frame, q, x=xyz_offset)
             u_task[:3] = xyz - target[:3]
 
-        # calculate orientation error if orientation is being controlled
+        # if orientation is being controlled
         if np.sum(self.ctrlr_dof[3:]) > 0:
+            u_task[3:] = self._calc_orientation_forces(target[3:], q)
 
-            # Method 1 --------------------------------------------------------
-            # transform the orientation target into a quaternion
-            q_d = transformations.unit_vector(
-                transformations.quaternion_from_euler(
-                    target[3], target[4], target[5], axes='rxyz'))
-            # get the quaternion for the end effector
-            q_e = transformations.unit_vector(
-                transformations.quaternion_from_matrix(
-                    self.robot_config.R('EE', q)))
-            # TODO: make sure that separate orientation and position gains
-            # work with velocity limiting and without
-            q_r = transformations.quaternion_multiply(
-                q_d, transformations.quaternion_conjugate(q_e))
-            u_task[3:] = -self.ko / self.kp * q_r[1:] * np.sign(q_r[0])
-
-            # # Method 2 --------------------------------------------------------
-            # # From (Caccavale et al, 1997) Section IV Quaternion feedback -----
-            # # get rotation matrix for the end effector orientation
-            # R_e = self.robot_config.R('EE', q)
-            # # get rotation matrix for the target orientation
-            # R_d = transformations.euler_matrix(
-            #     target[3], target[4], target[5], axes='rxyz')[:3, :3]
-            # R_ed = np.dot(R_e.T, R_d)  # eq 24
-            # q_ed = transformations.unit_vector(
-            #     transformations.quaternion_from_matrix(R_ed))
-            # u_task[3:] = -self.ko / self.kp * np.dot(R_e, q_ed[1:])  # eq 34
+        # task space integrated error term
+        if self.ki != 0:
+            self.integrated_error -= u_task
```

Comment:
why do we subtract u_task from the integrated error?

Reply:
fixed up signs